keps/sig-network/2004-topology-aware-routing/README.md

```diff
@@ -0,0 +1,553 @@
+provides backward compatibility for users. It represents the existing approach
+that is already used, no special mechanism will be added. Traffic will be
+evenly distributed among endpoints behind a Service.
+- `Require`: EndpointSlice controller will conduct a simple but aggressive
+routing mechanism. It will group endpoints in a zone to an EndpointSlice Group
+that only receives traffic from that zone. `Require` requires users to ensure
+that a sufficient number of endpoints for a Service exist in each zone.
+Services using this approach will only be reachable from zones that contain
+endpoints for the Service.
+- `Prefer`: EndpointSlice controller will conduct a more conservative routing
+mechanism. It will keep traffic in zone as long as each zone has a sufficient
+number of endpoints. If a zone does not have a sufficient number of endpoints,
+traffic will be routed to other zones.
+
+This same pattern could be used for topology aware routing at node and region
+level. We could add new annotations `endpointslice.kubernetes.io/same-node` and
+`endpointslice.kubernetes.io/same-region` for node and region respectively with
+the same values `Require` and `Prefer`. With all these three levels combined
+together, their individual value could be influenced by values of others, i.e.
+`Require` node would also `Require` zone and region.
+
+
+### Notes/Constraints/Caveats
+
+In the future, we may introduce another parameter for the `Prefer` approach
+that would allow users to specify a starting threshold. Starting threshold
+defines the starting point for `Prefer` (more detailed description below). We
+currently set the starting threshold to 3x the number of zones based on results
+from our test dataset for safer distribution results.
+
+### Risks and Mitigations
+
+- When users opt-in to the `Require` routing scheme, it is their responsibility
+to make sure the workloads are balanced.
+- In a scenario where all traffic originates from a single zone, with `Require`
+and `Prefer` options, there is a chance that endpoints in that zone will be
+overloaded while endpoints in other zones receive little to no traffic.
+
+## Design Details
+
+### Assumptions
+
+- Incoming traffic is proportional to the number of nodes in a zone. Although
+this is an imperfect metric, it is the best available way of predicting how
+much traffic will be received in a zone. We could alternatively consider basing
+this on the number of cores in a zone.
+- Service capacity is proportional to the number of endpoints in a zone. This
+assumes that each endpoint has equivalent capacity. Although this is not always
+true, it usually is. We can explore ways to deal with variable capacity
+endpoints in the future.
+
+### Kube-Proxy
+
+Kube-Proxy will be updated to support EndpointSlice subsetting. Each instance
+will watch for EndpointSlices that have a label indicating they should be
+consumed by this zone. For more information, refer to the EndpointSlice
```

Comment:
can we expand on this? how will kube-proxy know when it has to install routes to other zones or no?

Reply:
I owe another KEP here. The general idea would be adding a label to EndpointSlices like `endpointslice.kubernetes.io/for-zone: zone-a`. Then kube-proxy could be configured to only subscribe to EndpointSlices with that label set to the same zone or not set at all.

Reply:
Yeah that is quite messy. I'd missed that there's no way to express `OR` with selectors. This combination of selectors could work but would be bad if it dramatically increases the number of informers in kube-proxy/number of connections to api server from each node.

This could be a compelling reason to add some kind of global label to slices that didn't apply to a zone or region. That would allow us to express all of this with a single selector but unfortunately would result in all EndpointSlices being updated whenever this upgrade happened. This would also require any EndpointSlice producers to be updated.

Reply:
Now that I've filed https://github.com/kubernetes/enhancements/pull/2031, I think this discussion will actually make more sense there. I ended up taking a similar approach to what @danwinship proposed above with that proposal with the exception of no support for node labels/delivery.